Revisions Controller: return single revision only when it matches the parent id

ramonjd opened the pull request:
This patch adds a condition to `WP_REST_Revisions_Controller->get_item` that checks the revision parent id against the supplied parent id. 

If it doesn't match, the method returns a 404.

This is to ensure that  `get_item` does not return a revision whose parent does not match the `parent` route fragment.

Test coverage added.

Run tests: `npm run test:php -- --filter WP_Test_REST_Revisions_Controller`

Commit 1a7ff6c75c79398bec9504b4e70d4f2f95e9c310 demonstrates that currently, `get_item()` can return the revision of different parent post of that same type.


Trac ticket: https://core.trac.wordpress.org/ticket/59875

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

ramonjd committed:
Adding some tests to check whether get_item can return a revision whose parent does not match the `parent` route fragment.

ramonjd committed:
Add a condition to get_item that checks the revision parent id against the supplied parent id.
If it doesn't match, return a 404
Updated test coverage

Update error message in class-wp-rest-revisions-controller.php

ramonjd commented on tests/phpunit/tests/rest-api/rest-revisions-controller.php:
```diff
@@ -8,24 +8,29 @@
-	private $revision_id2;
-	private $revision_3;
-	private $revision_id3;
+	private $revision_1_1;
```
I renamed all these variable to indicate parent-> revision relationship.

ramonjd commented on tests/phpunit/tests/rest-api/rest-revisions-controller.php:
```diff
@@ -222,37 +247,59 @@ public function test_get_item_no_permission() {
+	/**
+	 * @ticket 59875
+	 */
+	public function test_get_item_valid_parent_id() {
```
`test_get_item_valid_parent_id` and `test_get_item_invalid_parent_id` below are the new tests.

andrewserong commented on src/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php:
```diff
@@ -402,6 +403,15 @@ public function get_item( $request ) {
+			return new WP_Error(
+				'rest_post_invalid_parent',
+				/* translators: %s: A post revision id. */
+				sprintf( __( 'Parent post does not have a revision with id of "%s"' ), $revision->post_parent ),
```
The error message mentions the id as being the revision id, but the passed in value is the revision's `post_parent`. Should `$revision->ID` (or `$request['id']`) be passed in instead, or adjust the error message to a more generic `'Parent post does not have the requested revision id'`?

ramonjd replied:
That was an oversight, thanks @andrewserong! I'll update.

I was trying to decide whether to highlight the invalid parent (with respect to the supplied revision id) or the invalid revision id (with respects to its parent), and my brain exploded.

andrewserong replied:
Thanks for updating!

andrewserong reviewed (commented):
Thanks for following up with this fix! Looks like a good approach to me, just left a tiny comment about the value used in the error message.

danielbachhuber commented on tests/phpunit/tests/rest-api/rest-revisions-controller.php:
```diff
@@ -8,24 +8,29 @@
  * @group restapi
  */
 class WP_Test_REST_Revisions_Controller extends WP_Test_REST_Controller_Testcase {
-	protected static $post_id;
```
In the interest of minimizing churn, can we keep all of the existing variables named as they are?

I recognize `post_id_1` is easier to follow, but all of the churn makes the history of the file much more difficult to follow.

ramonjd replied:
Fair enough. I think it's still worth removing some of the cognitive load of grokking these variables, so I'll revert the change here and create another patch.

Thanks!

danielbachhuber replied:
Thanks! Lot easier to read this way.

danielbachhuber commented on src/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php:
```diff
@@ -402,6 +403,15 @@ public function get_item( $request ) {
 			return $revision;
 		}
 
+		if ( $parent->ID !== $revision->post_parent ) {
```
```suggestion
		if ( (int) $parent->ID !== (int) $revision->post_parent ) {
```

danielbachhuber commented on src/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php:
```diff
@@ -402,6 +403,15 @@ public function get_item( $request ) {
 		}
 
+		if ( $parent->ID !== $revision->post_parent ) {
+			return new WP_Error(
```
This seems fine from a code perspective. How could it explode for consumers of the API?

ramonjd replied:
It won't bode well for folks who, for whatever reason, probably unintentionally, are expecting values back from a mismatched combination of parentId and revisionId. 

The argument for dismissing this concern is that, in such as case, the URL is not representational of the resource, and so it is not RESTful. It's probably worth a dev note on the Make Core blog.

How it could explode in another scenarios? I'm not sure. We're verifying the existence of both parent and revision posts, and their post types. I suppose if the expected properties, ID and post_parent, are unavailable but that would be a deeper issue I believe.

danielbachhuber replied:
Yeah, let's just call it a bug and make sure we mention the change of behavior in a Dev Note

jonnynews replied:
I think this should be moved to the permission check.

danielbachhuber replied:
@jonnynews Can you explain further? I think the current location is appropriate because it's a data discrepancy check, not a security check. There are data discrepancy checks immediately above too.

jonnynews replied:
It's it a security check, as the current user might not have access.

ramonjd replied:
> as the current user might not have access

I agree prima facie with @danielbachhuber here that it's a data mismatch check, but I'd like to understand the concerns. Could you please elaborate a little more? 

Do you mean access to the revision itself? There's a test in `get_items_permissions_check()` to ensure the user has edit rights to the parent. 

Here we're verifying whether the revision matches the parent, not explicitly whether the user has access to the revision entity.

Unless you think we should add a `current_user_can` for `$revision->ID`  in `get_items_permissions_check()`?

ramonjd replied:
Before we go ahead and merge this, it'd be helpful to have some clarification on your comment if you have time @jonnynews 

Otherwise we can always do a follow up PR with any constructive feedback.

Thanks a lot!

spacedmonkey replied:
Sorry, this is my other account. Looking at the rest of the revision controller, it seems that other checks are like this. So I am going to dismiss.

ramonjd committed:
Coerce types

ramonjd committed:
Remove unused var in test

danielbachhuber commented on src/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php:
```diff
@@ -402,6 +403,15 @@ public function get_item( $request ) {
+			return new WP_Error(
+				'rest_post_invalid_parent',
+				/* translators: %s: A post revision id. */
+				sprintf( __( 'Parent post does not have a revision with id of "%s"' ), $request['id'] ),
```
```suggestion
				sprintf( __( 'Parent post does not have a revision with id of "%d"' ), $request['id'] ),
```

danielbachhuber commented on src/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php:
```diff
@@ -402,6 +403,15 @@ public function get_item( $request ) {
+		if ( (int) $parent->ID !== (int) $revision->post_parent ) {
+			return new WP_Error(
+				'rest_post_invalid_parent',
+				/* translators: %s: A post revision id. */
```
```suggestion
				/* translators: %d: A post revision id. */
```

danielbachhuber reviewed (approved):
Looks good! Thanks for your work on this, @ramonjd 

I'll let @adamsilverstein give it a final review.

ramonjd committed:
Apply suggestions from code review

string replacement type signifier %d

Co-authored-by: Daniel Bachhuber <daniel@bachhuber.co>

ramonjd commented:
Thanks, folks. I appreciate the testing and reviews!

adamsilverstein reviewed (approved):
Nice work!

jonnynews commented:
@ramonjd Has this change been tested with https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-template-revisions-controller.php controller that extends this controller?

ramonjd commented:
> Has this change been tested with https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-template-revisions-controller.php controller that extends this controller?

Good point. I can add a relevant unit test to the `WP_REST_Template_Revisions_Controller` test suite.

ramonjd commented:
~~I can confirm that we'll need to overwrite `get_item()` in `WP_REST_Template_Revisions_Controller` to account for the `wp_id` > `parent` relationship between template record and their revisions.~~

<img width="1398" alt="Screenshot 2023-11-28 at 1 56 16 pm" src="https://github.com/WordPress/wordpress-develop/assets/6458278/83ad9a56-ef09-48d4-85c2-7d7eff74c2fc">

~~I'll do this and add a test to this PR since it's related, and also check any other post types that extend this controller.~~ 

No. I was mixing up the results of `get_post` and `WP_REST_Template_Revisions_Controller`. 

See [below](https://github.com/WordPress/wordpress-develop/pull/5655#issuecomment-1829011113).

ramonjd commented:
> I can confirm that we'll need to overwrite get_item() in WP_REST_Template_Revisions_Controller to account for the wp_id > parent relationship between template record and their revisions.

Actually, withdraw that, I should have checked `WP_REST_Revisions_Controller::get_item` before posting.

The check in `get_item` that this PR introduces does so by comparing the results of `get_post` for both the parent and the revision.

The results of `get_post` is independent of the `WP_REST_Template_Revisions_Controller` which performs all the `wp_id` faffery: `get_post` will return integer IDs and parent IDs for template posts and revisions alike.

Regardless, `WP_REST_Template_Revisions_Controller` has [unit tests](https://github.com/ramonjd/wordpress-develop/blob/f8bf0daa1acfa68f2b421cfd6149fb9acccd41c6/tests/phpunit/tests/rest-api/wpRestTemplateRevisionsController.php#L299-L299) for `get_item`, which would have failed otherwise.

I've also double checked other registered post types, none of which, `wp_template` and `wp_template_part` notwithstanding, define a custom revisions_rest_controller_class.

I'd submit, therefore, that this PR is fine as it is.

Thanks for the reviews!

spacedmonkey reviewed (approved):
👍 This looks good to me.

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php:
```diff
@@ -402,6 +403,15 @@ public function get_item( $request ) {
+			return new WP_Error(
+				'rest_post_invalid_parent',
+				/* translators: %d: A post revision id. */
+				sprintf( __( 'Parent post does not have a revision with id of "%d"' ), $request['id'] ),
```
@ramonjd Should this message be overridden in `WP_REST_Template_Revisions_Controller`? 

https://github.com/WordPress/wordpress-develop/blob/c78e3fbce4c8af24b8feff47e6f40f9f7873a7db/src/wp-includes/rest-api/endpoints/class-wp-rest-template-revisions-controller.php#L169-L175

Similar messages there use the wording template.

spacedmonkey commented on tests/phpunit/tests/rest-api/rest-revisions-controller.php:
```diff
@@ -234,6 +256,28 @@ public function test_get_item_invalid_parent_post_type() {
 		$this->assertErrorResponse( 'rest_post_invalid_parent', $response, 404 );
 	}
 
+	/**
```
@ramonjd I would like these invalid tests to added to `wpRestTemplateRevisionsController.php`. I like testing unhappy paths as well as happy paths.

ramonjd replied:
Good call, thanks.

Done in ae49d86c2070ac1bc06461a52f52c1d5a42301bb

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php:
```diff
@@ -402,6 +403,15 @@ public function get_item( $request ) {
+			return new WP_Error(
+				'rest_post_invalid_parent',
+				/* translators: %d: A post revision id. */
+				sprintf( __( 'Parent post does not have a revision with id of "%d"' ), $request['id'] ),
```
@ramonjd Should this message be overridden in `WP_REST_Template_Revisions_Controller`? 

https://github.com/WordPress/wordpress-develop/blob/c78e3fbce4c8af24b8feff47e6f40f9f7873a7db/src/wp-includes/rest-api/endpoints/class-wp-rest-template-revisions-controller.php#L169-L175

Similar messages there use the wording template.

ramonjd replied:
Good catch. I've copied/pasted the same error code as the `WP_Error()` in the the `get_parent` method: `rest_post_invalid_parent`.

I'll update the mismatch error to something more unique and descriptive.

spacedmonkey replied:
@ramonjd I didn't mean the error code, I meant the message. It should say something like 
```
sprintf( __( 'This template does not have a revision with id of "%d"' ), $request['id'] ),
```

This would mean either overriding the `get_item` method in template revisions controller or moving this logic to it's own method.

ramonjd replied:
Oh, apologies. Yeah, I guess "post" in the string isn't quite accurate in that context

`sprintf( __( 'Parent post does not have a revision with id of "%d"' ), $request['id'] )`

Rather than create more code in `class-wp-rest-template-revisions-controller.php`, what would you say about genericizing the message? E.g., 

`sprintf( __( 'Parent does not have a revision with id of "%d"' ), $request['id'] )`

I'll push a commit with that, but happy to extend `class-wp-rest-template-revisions-controller.php` as well.

ramonjd replied:
> This would mean either overriding the get_item method in template revisions controller or moving this logic to it's own method.

Now that I've had my coffee, I think I do prefer the messaging of `"'This template does not have a revision with id of "%d"'`.

Perhaps moving the mismatch logic to a single method might be more testable. I'll do that. 👍🏻

ramonjd replied:
The coffee has worn off, and I went with genericizing the error message further to avoid creating more code to maintain. 

I added a test to ensure the correct parent ID is returned in the error message.

```php
return new WP_Error(
    'rest_revision_parent_id_mismatch',
    /* translators: %d: A post id. */
    sprintf( __( 'The revision does not belong to the specified parent with id of "%d"' ), $parent->ID ),
    array( 'status' => 404 )
);
```

ramonjd commented:
Thanks a lot for the feedback @spacedmonkey! 🙇🏻 I'll get it done today or tomorrow.

ramonjd committed:
Changed duped error code
Adding parent id revision parent_id mismatch invalid test to the templates revisions controller tests

Remove specific mention of "post" to discern between other post types, e.g., "template"

ramonjd committed:
Genericize error message and create a test to ensure the correct parent ID is returned in the error message.

ramonjd committed:
Remove specific mention of "post" to discern between other post types, e.g., "template"

adamsilverstein reviewed (approved):
Excellent, thanks for all the tests!

tellthemachines commented:
Committed in [r57222](https://core.trac.wordpress.org/changeset/57222).
